client/www/lib/hooks/useLocalStorage.tsx

```diff
@@ -0,0 +1,57 @@
+import { useCallback, useRef, useSyncExternalStore } from 'react';
+
+function getSnapshot<T>(k: string): T | undefined {
+  if (typeof window == 'undefined') return;
+  let v = window.localStorage.getItem(k);
+  if (!v) return;
+  try {
+    return JSON.parse(v);
+  } catch (e) { }
+}
+
+function getServerSnapshot(k: string): undefined {
+  return;
+}
+
+function setItem<T>(k: string, v: T | undefined) {
+  if (typeof window == 'undefined') {
+    throw new Error('useLocalStorage/setState needs to run on the client');
+  }
+  const stringified = JSON.stringify(v);
+  window.localStorage.setItem(k, stringified);
+  // localStorage.setItem does not dispatch events to the current
+  window.dispatchEvent(
+    new StorageEvent('storage', { key: k, newValue: stringified })
+  );
+}
+
+export function useLocalStorage<T = {}>(
+  k: string
+): [T | undefined, (v: T | undefined) => void] {
+  const snapshotRef = useRef<T | undefined>(getSnapshot<T>(k));
+  const subscribe = useCallback((cb: Function) => {
+    const listener = () => {
+      snapshotRef.current = getSnapshot<T>(k);
+      cb();
+    };
+    window.addEventListener('storage', listener);
+    return () => {
+      window.removeEventListener('storage', listener);
+    };
+  }, []);
+  const state = useSyncExternalStore<T | undefined>(
+    subscribe,
+    () => snapshotRef.current,
+    () => getServerSnapshot(k)
+  );
+  return [state, (v: T | undefined) => setItem<T>(k, v)];
+}
+
+export function useLocalStorageWithDefaultValue<T>(
```

Comment:
I had to make a separate function for defaultValue. Otherwise the types get _really_ hairy. I had a hard time getting TS to understand that _if_ a defaultValue is provided, then the result should refine to `T`

Reply:
Update: went ahead and just made defaultValue required. Maps similar to useState